Audit ChatMessageComposer Attachments

skorepak opened the pull request:
## Description of the pull request
Audit and merge Image and Documents Attachments in the `ChatMessageComposer`

### I have
- Moved the CollectionView data source and delegate from collection view itself into separated ViewController and used this as a child view controller in the `ChatMessageComposerVC`
- Merged most of the views for Images and Documents and IMO simplified the code in exchange for two switches based on attachment type
- Renamed Documents to Files to match the name of attachment type
- Removed a few magic constants

## Links
https://stream-io.atlassian.net/browse/CIS-749

nuno-vieira commented on Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:
```diff
@@ -221,6 +226,7 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
                 messageId: threadParentMessage.id
             )
             
+            #warning("Possible issue (typo?): 'attachments: attachments + attachments'")
             messageController?.createNewReply(
                 text: text,
                 pinning: nil,
```
Yes that is a typo, I've fixed that on my branch too 👍

nuno-vieira commented:
> ### I have not (hence the draft)
> * Wrote tests yet

@skorepak BTW, don't bother writing the Audit tests yet. The goal is to refactor all the composer view, and after that writing the audits for all the new structure 👍

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentCollectionViewCell.swift:
```diff
@@ -25,7 +25,7 @@ open class _AttachmentCollectionViewCell<ExtraData: ExtraDataTypes>: _Collection
     public var discardAction: (() -> Void)?
     
-    public private(set) var attachmentView: AttachmentView? = nil
+    public private(set) var attachmentView: AttachmentView?
```
I think this should be open 🤔 I can see we have `public` and `open` in some components, but in order to change some child components in specific views, without changing them everywhere with the `components` config, all views should be `open private(set)` right? @VojtaStavik I think this is written in the Audits Documents, but maybe is not clear enough.

skorepak replied:
The idea was that `attachmentView` should be created with `func attachmentView(for attachmentType: AttachmentType) -> AttachmentView?`  because the updateContent function is using it. Therefore not open, but I can open it. Want to add something @VojtaStavik ?

nuno-vieira commented on Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:
```diff
@@ -245,27 +248,38 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
+        // The UI doesn't support mix of image and file attachments so we are limiting this option.
+        // Files in the message composer are scrolling vertically and images horizontally.
+        // There is no techical limitation for multiple attachment types.
+        if urlAttachments.isEmpty {
+            selectedAttachmentType = nil
             present(actionSheet, animated: true, completion: nil)
-        case .media:
-            present(imagePicker, animated: true, completion: nil)
-        case .documents:
-            present(documentPicker, animated: true, completion: nil)
+        } else if let attachmentType = selectedAttachmentType {
+            switch attachmentType {
+            case .image:
+                present(imagePicker, animated: true, completion: nil)
+            case .file:
+                present(filePicker, animated: true, completion: nil)
+            default:
+                log.error("Error in handling `selectedAttachmentType`. Unexpected attachment type.")
+            }
+        } else {
+            log.error("Error in handling `selectedAttachmentType`. Missing attachment type.")
```
Why are we limiting this actually? 🤔 Is this an LLC Support limitation?

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/FileAttachmentView.swift:
```diff
@@ -5,11 +5,16 @@
+    /// A closure handler that is called when the discard button of the attachment is clicked
     public var discardButtonHandler: (() -> Void)?
     
     // MARK: - Subviews
```
same as above

skorepak commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentPreviews/DocumentAttachmentPreview.swift:
```diff
@@ -0,0 +1,24 @@
+import UIKit
+
+/// The image attachment preview information.
+public struct DocumentAttachmentPreview: AttachmentPreview {
+    /// The image attachement type
+    public var type: AttachmentType = .file
+    /// The image preview.
+    public let image: UIImage
+    /// The name of the image attachment.
+    public let name: String?
+    /// The size of the attachment.
+    public let size: Int64
+
+    public init(image: UIImage, name: String?, size: Int64) {
+        self.image = image
+        self.name = name
+        self.size = size
+    }
+}
```
Missed this one, thank you! :)

skorepak commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/FileAttachmentView.swift:
```diff
@@ -24,7 +29,8 @@ open class _ChatDocumentAttachmentView<ExtraData: ExtraDataTypes>: _View, ThemeP
+    open private(set) lazy var discardButton: UIButton = UIButton()
         .withoutAutoresizingMaskConstraints
 
     public private(set) lazy var fileNameAndSizeStack: UIStackView = {
```
Right! I didn't touch those views where if I didn't have to, but will update this :)

skorepak commented on StreamChat.xcodeproj/project.pbxproj:
```diff
@@ -6458,7 +6468,7 @@
+				PROVISIONING_PROFILE_SPECIFIER = "";
 				SWIFT_VERSION = 5.0;
 				TARGETED_DEVICE_FAMILY = "1,2";
 			};
```
Probably not, will reset 👍🏻

VojtaStavik commented on Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:
```diff
@@ -343,63 +359,37 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
-    
-    open var documentAttachments: [URL] = [] {
+    open var urlAttachments: [URL] = [] {
         didSet {
-            didUpdateDocumentAttachments()
+            didUpdateAttachments()
         }
     }
```
Do we actually need this indirection? Why not use `var attachments: [AttachmentEnvelope]` directly? It will make it also possible to add custom attachments there by customers.

skorepak replied:
My understanding is that we work with URLs to save resources until necessary, I don't see a problem with using `[AttachmentEnvelope]` directly other than it's consuming more memory and taking some CPU time. 
If I'm thinking about it, every change is directly visible in the Attachments view anyway, so in the end, it doesn't have to more expensive.

skorepak replied:
Done

VojtaStavik commented on Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:
```diff
@@ -343,63 +359,37 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
+    open var selectedAttachmentType: AttachmentType?
+
+    open func didUpdateAttachments() {
+        attachmentsViewController.content = attachments.compactMap {
+            switch $0.payload {
+            case let payload as AttachmentImagePayload:
+                guard let preview = UIImage(contentsOfFile: payload.imagePreviewURL.path) else { return nil }
+                return ImageAttachmentPreview(image: preview)
+                
+            case let payload as AttachmentFilePayload:
+                let filePreview = $0.localFileURL.flatMap { appearance.images.documentPreviews[$0.pathExtension] }
+                    ?? appearance.images.fileFallback
+                return FileAttachmentPreview(image: filePreview, name: payload.title, size: payload.file.size)
+
+            default:
+                log.error("Unexpected attachment payload type.")
+                return nil
+            }
         }
```
What if instead of this created something like:
```
protocol AttachmentPreview { 
  func previewView() -> UIView // or UIView? or even AttachmentPreviewView?
}
```
and said that `open var attachments: [AttachmentEnvelope & AttachmentPreview]`? Customers will be able to implement this protocol and provide their own implementation for this (i.e. map preview, etc.).

skorepak replied:
Done, please check

VojtaStavik commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentCollectionViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+import StreamChat
+import UIKit
+
+public protocol AttachmentView: _View {
+    var content: AttachmentPreview? { get set }
+    var discardButtonHandler: (() -> Void)? { get set }
+}
```
nit: `AttachmentPreviewView`? Rather than from `_View` it would be better if it inherited from `UIView` to make it possible to use custom attachments

VojtaStavik commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentCollectionViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+        attachmentView?.removeFromSuperview()
+    }
+    
+    open func attachmentView(for attachmentType: AttachmentType) -> AttachmentView? {
+        let type: AttachmentView.Type
+        
+        switch attachmentType {
+        case .image:
+            type = components.messageComposer.imageAttachmentView
+
+        case .file:
+            type = components.messageComposer.fileAttachmentView
+
+        default:
+            log.error("Unsuported attachment type: \(attachmentType)")
+            return nil
+        }
+        
+        // Cell reuse optimalisation
+        // Attachment view is already the right type. Don't create it again, just prepare it for reuse.
+        if attachmentView?.isKind(of: type) == true {
+            attachmentView?.content = nil
+            attachmentView?.discardButtonHandler = nil
+            return attachmentView
+        }
+        
+        let result: AttachmentView = type.init()
+        result.translatesAutoresizingMaskIntoConstraints = false
+        return result
+    }
```
Wouldn't it be better if this happened on the controller level?

skorepak replied:
Now it's happening on the attachment level, as per discussion

VojtaStavik reviewed (commented):
Nice 👍  It's a bit hard to review like this but I left some high-level comments. Maybe let's discuss it more in Slack?

skorepak commented:
Additionally, I have:
- replaced the collection view with container stack view
-  removed attachmentView type decision on the CollectionViewCell level. 
-  extended the AttachmentPreview with a `previewView()` function to make Attachment responsible for providing the correct view for itself. This will make creating custom attachments much easier. 

There is one tiny layout issue that I'm still working on, it should not change the code much.

evsaev commented on Sources/StreamChat/Models/Attachments/ChatMessageAttachmentEnvelope.swift:
```diff
@@ -23,12 +23,12 @@ public extension ChatMessageAttachmentEnvelope {
     }
 
-    init(localFileURL: URL) throws {
+    init(localFileURL: URL, attachmentType: AttachmentType) throws {
```
@skorepak do we really need this? 🤔 This API allows to give a custom attachment type but the attachment can't be updated when the file is uploaded since the `type` is different from `file/image`

evsaev replied:
basically wondering what the use case is 🙂 Sending an image as a file attachment or ..? If we are sure the type has to be passed explicitly other parts of attachment uploading flow (`AttachmentDTO.update(uploadedFileURL:)`) should be updated accordingly

skorepak replied:
We need this to distinguish between the image uploaded from file picker vs from image picker. Without this, when the image was picked through a file picker, we showed the image as an image, even tho it was picked as a file.

evsaev commented on Sources/StreamChatTestTools/Models/Attachments/ChatMessageAttachmentEnvelope_Mock.swift:
```diff
@@ -5,7 +5,7 @@
 @testable import StreamChat
 
-public extension ChatMessageAttachmentEnvelope {
+public extension AnyAttachmentPayload {
```
nit: `ChatMessageAttachmentEnvelope_Mock` -> `AnyAttachmentPayload_Mock`

VojtaStavik replied:
fixed in 2c405cf, thanks!

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentPlaceholder.swift:
```diff
@@ -0,0 +1,42 @@
+import StreamChat
+import UIKit
+
+/// Default provider that is used when AttachmentPreviewProvider is not implemented for custom attachment payload.
+public struct AttachmentPreviewPlaceholderProvider: AttachmentPreviewProvider {
+    public func previewView<ExtraData: ExtraDataTypes>(components: _Components<ExtraData>) -> UIView {
+        AttachmentPlaceholderView().withoutAutoresizingMaskConstraints
+    }
+}
```
nit: I find that naming actual "Default" on default implementations it is easier to understand what is this implementation about
```suggestion
/// Default provider that is used when AttachmentPreviewProvider is not implemented for custom attachment payload.
public struct DefaultAttachmentPreviewProvider: AttachmentPreviewProvider {
    public func previewView<ExtraData: ExtraDataTypes>(components: _Components<ExtraData>) -> UIView {
        AttachmentPlaceholderView().withoutAutoresizingMaskConstraints
    }
}
```

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentPreviewProvider.swift:
```diff
@@ -0,0 +1,42 @@
+    /// The preferred axis to be used for attachment previews in attachments view.
+    static var preferredAxis: NSLayoutConstraint.Axis { get }
+}
+
+public extension AttachmentPreviewProvider {
+    static var preferredAxis: NSLayoutConstraint.Axis { .horizontal }
+}
```
Why do we need a default implementation here? It's easier for who is implementing this protocol to decide the axis.

VojtaStavik replied:
fixed, thanks!

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/FileAttachmentView.swift:
```diff
@@ -5,15 +5,31 @@
-    // MARK: - Properties
+/// A view that displays the file attachment.
+open class FileAttachmentView: _View, AppearanceProvider {
+    open var height: CGFloat = 54
```
On the other views, we are not making these properties available. The values are hardcoded in the `setupLayout` and if the customer wants to change it need to override the `setupLayout()`. Because all our constraints are with 900 priority, the customer constraint will override ours.

VojtaStavik replied:
-> slack

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/ImageAttachmentView.swift:
```diff
@@ -0,0 +1,47 @@
+
+/// A view that displays the image attachment.
+open class ImageAttachmentView: _View, AppearanceProvider {
+    open var width: CGFloat = 100
+    open var height: CGFloat = 100
```
Same as above ☝️

nuno-vieira commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentsPreviewVC.swift:
```diff
@@ -0,0 +1,137 @@
+    override open func updateContent() {
+        super.updateContent()
+        
+        switch stackViewAxis {
+        case .horizontal:
+            setupHorizontalStackView()
+            
+        case .vertical:
+            setupVerticalStackView()
+            
+        @unknown default:
+            break
+        }
```
We probably could freeze this enum 🤔

VojtaStavik replied:
@nuno-vieira we don't own it 😛

nuno-vieira replied:
Ahh I see, this is the axis from Apple 😄

DominikBucher12 commented on Sources/StreamChatUI/Utils/UIViewController+Extensions.swift:
```diff
@@ -19,6 +19,14 @@ extension UIViewController {
+        child.willMove(toParent: self)
+        addChild(child)
+        child.view.translatesAutoresizingMaskIntoConstraints = false
+        superview.embed(child.view)
```
nit, I tought we are surpressing `embed` function? :D

VojtaStavik replied:
We will, soon :)

DominikBucher12 commented on Sources/StreamChatUI/CommonViews/Attachments/DefaultAttachmentPreviewProvider.swift:
```diff
@@ -0,0 +1,45 @@
+        super.layoutSubviews()
+
+        borderLayer.frame = bounds
+        borderLayer.path = UIBezierPath(roundedRect: bounds, cornerRadius: 16).cgPath
```
This is probably overthinking 😅 but if we want to make our API even more convenient, we could extract some cornerRadius types and create some extension on float that we would type: cornerRadius: .halfOfheight(of: view) or cornerRadius: .medium Currently if you take a look into the project there is really lot of different corner-radiuses and it feels it's not unified across views... Maybe we can discuss this with designers later to make this nicer and for our customers not to override anything when wanting to unify those..

 But this is not important right now

VojtaStavik replied:
Yeah, good point 👍 Feel free to create a JIRA ticket for it

DominikBucher12 commented on Sources/StreamChatUI/Composer/ComposerVC.swift:
```diff
@@ -319,6 +319,11 @@ open class _ComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
         )
+        
+        attachmentsVC.content = content.attachments.map {
+            ($0.payload as? AttachmentPreviewProvider) ?? DefaultAttachmentPreviewProvider()
```
shouldn't the absence of payload be somehow logged at least as `info`?

VojtaStavik replied:
fixed, thanks!

DominikBucher12 commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/FileAttachmentView.swift:
```diff
@@ -0,0 +1,106 @@
+        super.setUpAppearance()
+        
+        backgroundColor = appearance.colorPalette.background
+        layer.cornerRadius = 15
```
As previous comment about unifying the corner radius stated, this would be nice + it feels strange to me that this is `15` and previews are `16`...

VojtaStavik replied:
fixed, thanks!

DominikBucher12 commented on Sources/StreamChatUI/CommonViews/Attachments/AttachmentViews/ImageAttachmentView.swift:
```diff
@@ -0,0 +1,47 @@
+        super.setUpAppearance()
+
+        layer.masksToBounds = true
+        layer.cornerRadius = 15
```
again 15 vs 16 🤔

VojtaStavik replied:
fixed, thanks!
